File: 142.Linked-List-Cycle-II.go
// https://leetcode.com/problems/linked-list-cycle-ii/
//
// algorithms
// Medium (30.29%)
// Total Accepted:    187,091
// Total Submissions: 617,612
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	first, two := head.Next, head.Next.Next
	hasCycle := false

	for first != nil && two != nil {
		if first == two {
			hasCycle = true
			break
		}
		first = first.Next
		if two.Next == nil {
			break
		}
		two = two.Next.Next
	}

	if !hasCycle {
		return nil
	}

	first = first.Next
	length := 1
	for first != two {
		first = first.Next
		length++
	}

	first, two = head, head
	for i := 0; i < length; i++ {
		first = first.Next
	}

	for first != two {
		first = first.Next
		two = two.Next
	}

	return first
}

File: 142.Linked-List-Cycle-II.java
// https://leetcode.com/problems/linked-list-cycle-ii/
//
// algorithms
// Medium (31.93%)
// Total Accepted:    212,281
// Total Submissions: 664,890
// beats 100.0% of java submissions


/**
 * Definition for singly-linked list. class ListNode { int val; ListNode next;
 * ListNode(int x) { val = x; next = null; } }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }

        ListNode st = head, nd = head.next;

        while (nd != null && st != nd) {
            st = st.next;
            nd = nd.next;
            if (nd != null) {
                nd = nd.next;
            }
        }

        if (nd == null) {
            return null;
        }

        int cycleLen = 1;
        st = nd.next;
        while (st != nd) {
            cycleLen++;
            st = st.next;
        }

        st = head;
        nd = head;
        for (int i = 0; i < cycleLen; i++) {
            st = st.next;
        }

        while (st != nd) {
            st = st.next;
            nd = nd.next;
        }

        return st;
    }
}